### ISSUE TITLE
[DEPR] Branding without design tokens

### ISSUE THREAD

**Issue by brian-smith-tcril** [openedx/brand-openedx#23]
### RFC Start Date

2024-07-19

### Target Plan Accepted Date

2024-08-02

### Target Transition Unblocked Date

2025-05-22

### Earliest Breaking Changes Unblocked Date

2025-06-18

### Earliest Open edX Named Release Without This Functionality

Ulmo - 2025-05

### Rationale

https://github.com/openedx/paragon/blob/master/docs/decisions/0019-scaling-styles-with-design-tokens.rst
https://github.com/openedx/paragon/tree/alpha?tab=readme-ov-file#design-tokens

### Removal

from https://github.com/openedx/paragon/blob/master/docs/decisions/0019-scaling-styles-with-design-tokens.rst#implications-for-theme-authors-eg-edxbrand-packages

> Currently, theme authors largely modify SCSS variables from core Paragon by creating a `_variables.scss` file and importing it after the core Paragon SCSS styles in consuming applications (e.g., micro-frontends). Doing so, SCSS will override the original variables' values defined by core Paragon with the new SCSS from the `@edx/brand` theme.

### Replacement

* See Also: [Implications for Theme Authors](https://github.com/openedx/paragon/blob/master/docs/decisions/0019-scaling-styles-with-design-tokens.rst#implications-for-theme-authors-eg-edxbrand-packages
)
> With design tokens, theme authors will instead override core Paragon tokens by defining their own JSON tokens that get deep merged alongside the core Paragon tokens, thus overriding any tokens that were defined by the theme author.

* Note: Existing themes packages will not work once this change has landed, see the Migration section for how to migrate.

### Deprecation

There is no plan to mark the code for deprecation. This will be a one time switch.

### Migration

There are plans to create a how-to document for migrating existing brand packages to design tokens, as well as discussions being had around building out automated tooling to assist in the process.

As of writing, the following information is available:
* **[Theming With Design Tokens](https://github.com/openedx/paragon/tree/release-23.x?tab=readme-ov-file#theming-with-design-tokens)**
* Compiling CSS from design tokens ([Paragon readme](https://github.com/openedx/paragon/tree/alpha?tab=readme-ov-file#design-tokens))
* Implications for theme authors ([Paragon ADR](https://github.com/openedx/paragon/blob/alpha/docs/decisions/0019-scaling-styles-with-design-tokens.rst#implications-for-theme-authors-eg-edxbrand-packages))
* Comparisons between design tokens ([Paragon design tokens directory](https://github.com/openedx/paragon/tree/alpha/tokens)) and pre-design tokens Paragon styles ([Paragon scss directory](https://github.com/openedx/paragon/tree/master/scss/core))
* Comparisons between [brand-edx.org design tokens](https://github.com/edx/brand-edx.org/tree/alpha/tokens/src) and [brand-edx.org scss](https://github.com/edx/brand-edx.org/tree/master/paragon).
* https://github.com/openedx/brand-openedx/pull/26

### Task List

There are 3 main phases:

1.  frontend-platform documentation (step by step guide to migration)
1. mfe upgrades - upgrade every MFE to use design tokens
1. tooling for tutor - tooling for building and hosting CSS variables

Additionally work has already been done to upgrade paragon to support design tokens, and frontend-app-discussions has been built on a pilot branch to use design tokens, proving out the system.

These MFEs are considered out of scope for the conversion:

```
openedx/frontend-app-admin-portal
openedx/frontend-app-enterprise-public-catalog
openedx/frontend-app-learner-portal-enterprise
openedx/frontend-enterprise
openedx/frontend-app-learner-portal-programs
openedx/frontend-app-ecommerce
openedx/frontend-app-payment
openedx/frontend-component-ai-translations
openedx/frontend-app-shell
```

**Comment by brian-smith-tcril** [comment]
The comment period has ended. In Teak pre-design-tokens brand packages will no longer be supported.

**Comment by robrap** [comment]
[inform/request] This ticket was accepted under the old DEPR process. The [updated Process States in OEP-21](https://open-edx-proposals.readthedocs.io/en/latest/processes/oep-0021-proc-deprecation.html#process-states) help add clarity to communications and timing regarding the introduction of breaking changes. This ticket must move to Transition Unblocked before moving to Breaking Changes Unblocked. The default minimum time to wait in Transition Unblocked is 1 month, so early deployers have time to prepare. If you wish to make breaking changes in less than 1 month, please initiate discussions to negotiate a date as soon as possible. Thank you!

**Comment by sarina** [comment]
Pulling in @adamstankiewicz here. I believe 2U is prepared already for this transition. I have seen the Elm theme working with Design Tokens.

**Comment by brian-smith-tcril** [comment]
The [Transition Unblocked documentation](https://open-edx-proposals.readthedocs.io/en/latest/processes/oep-0021-proc-deprecation.html#transition-unblocked) states

> Both old and new implementations operate in parallel

I'd like clarification on this. I believe the `master-design-tokens` branches of MFEs satisfy this requirement, as site operators can deploy those branches instead of deploying from the MFE `master` branches before we (simultaneously) land all of the `master-design-tokens` branches on the MFE `master` branches. 

This doesn't quite fit with

> Feature flags/toggles control which implementation is active

but I believe in this scenario configuring a deploy process to use different branches of MFEs effectively provides site operators the ability to "control which implementation is active."

**Comment by robrap** [comment]
@brian-smith-tcril:
1. I'll try to get someone on this ticket to ensure your understanding that those branches are enough. My understanding is that is the best that can be done for this.
2. Let's assume it does, and you are ready for Transition Unblocked. Do you have a clear date in mind for the breaking changes? Is that in less than a month or more than a month?
Thank you.

**Comment by sarina** [comment]
Can we wait until @adamstankiewicz weighs in here? If 2U is ready for the transition (as I think they are), and the branches have already been ready for around 2 weeks, and we plan to merge this within 2 weeks, that's about a month.

**Comment by robrap** [comment]
Note that the one month is for early deployers, and as far as we know right now, that is 2U and MIT.

**Comment by sarina** [comment]
We have been communicating with 2U and MIT about these changes for months :)

**Comment by robrap** [comment]
@sarina - I'm glad there has been a lot of communication. At the same time, this was also true for the footer slots, and yet we ran into trouble that I'm looking to avoid through the following:
1. Ensuring that planned dates are captured on the DEPR, so anyone following the DEPR will be clear. From the updated DEPR template, this would be:
> Earliest Breaking Changes Unblocked Date*
>
> When is the earliest date you may make breaking changes? A good default is the later of 6 months after the RFC Start Date, or 1 month after the Target Transition Unblocked Date. This date can be negotiated as part of the RFC process. See [OEP-21](https://open-edx-proposals.readthedocs.io/en/latest/processes/oep-0021-proc-deprecation.html) for more details.
2. Ensuring that the ticket captures, or points to docs for operators to complete the transition from old to new. I don't know if all 2U MFE owners are clear about the need to deploy off the `master-design-tokens` branches, or if they are planning on just handling the breakage when it comes in, or if they have some other plan, or if some are totally in the dark.

I know communication within 2U is not your problem, but ensuring that the ticket captures what is needed, or links to what is needed, will help me ensure I can communicate this out. Thank you.

**Comment by sarina** [comment]
I am going to move this to Transition Unblocked because we have had the master design tokens branches already available for consumption and this has been communicated to early deployers already.

I think @adamstankiewicz and @brian-smith-tcril are cooperating on the final date to merge in the DT branches to master/main, I will let them comment on their target date.

**Comment by robrap** [comment]
@sarina: Ok, but @feanil was added to this ticket as the coordinator who could follow through all the DEPR WG changes. If you wish to move the ticket, could you also update the PR Description to match the updated DEPR template? Also, once you have a date, can update the board field "Review After Date" with that date? Thanks!

**Comment by sarina** [comment]
@feanil is out sick this week, I will coordinate with him on this ticket when he returns.

**Comment by adamstankiewicz** [comment]
> Pulling in [@adamstankiewicz](https://github.com/adamstankiewicz) here. I believe 2U is prepared already for this transition. I have seen the Elm theme working with Design Tokens.

2U (primarily myself and @MaxFrank13 at the moment, but a couple others included) are aware of these changes and the implications for MFE owners. Max and I have recently been actively spending significant portions of our regular team's sprint capacity for the past few weeks to get the relevant 2U-owned dependencies ready for this migration. We are fairly close with the remaining implementation work on the upstream dependencies, but there are some final steps remaining that need to be worked out within 2U:
* Finalize `@edx/elm-theme` itself 
    * We had to recently devise a [new theming mechanism](https://github.com/openedx/paragon/pull/3598) to allow brand packages to override default button variants used in MFE code without modifying `openedx` MFE code. `@edx/elm-theme` itself still needs to adopt this new theming mechanism for (at least) 2 components.
    * Handful of identified minor design discrepancies between actual theme vs. the Elm Theme Figma; *most are non-blocking.*
    * A non-engineer design review for `@edx/elm-theme` would be nice; QA largely been Max and I as there appears to be no 2U designers actively involved with Paragon anymore. We are still hoping to find an internal designer familiar with elm-theme who can help at least click around the Paragon documentation website with the `@edx/elm-theme` applied to try to find any brand issues across the components before we make release announcements to MFE-owning squads (TBD). 
* Get our [`header-edx`](https://github.com/edx/frontend-component-header-edx/pull/663) and [`footer-edx`](https://github.com/edx/frontend-component-footer-edx/pull/480) PRs merged (should be complete by end of this week).
* Formalize an internal release plan and coordinate with MFE-owning squads to merge a PR to bump their header/footer/brand versions, verify their MFEs on stage, and deploy from there. 
    * Max and I have not yet gotten to draft/announce this as a formal plan yet, given our focus has been on making sure the relevant 2U dependencies are in a good spot first and foremost, given our limited capacity.
        * @robrap I plan to help facilitate a release plan within 2U once our `@edx/elm-theme` and other dependencies are ready to use (prioritizing implementation > release planning, atm). If you'd like to help support this coordination to get a head start on it, that'd be great! Let me know if you want to connect.

---

> I don't know if all 2U MFE owners are clear about the need to deploy off the master-design-tokens branches, or if they are planning on just handling the breakage when it comes in, or if they have some other plan, or if some are totally in the dark.

@robrap The primary intent for the `master-design-tokens` branches is to be able to get all Tutor-supported MFEs ready for Paragon v23 so they can all get merged into `master` at roughly the same time (coordinated release). 2U specifically did not have any plans to try to deploy from `master-design-tokens` before it merges to `master`.

2U MFE owners will need to deal with the breakage when it comes in, though, I'm hoping to help facilitate a internal release plan where I'll create the small PRs to bump the appropriate package versions for the relevant MFEs so 2U MFE owners should only have to review a PR, merge, verify nothing in their MFEs' core flows looks obviously broken/poorly styled, and deploy.

---

FWIW, there is some relevant MFE-specific documentation that might be finalized or moved/referenced in an appropriate place so it's more discoverable, for example:
* [Migrating MFEs to Paragon design tokens and CSS variables](https://openedx.atlassian.net/wiki/spaces/BPL/pages/3770744958/). Meant to serve a how-to guide for consuming MFEs. Most of these migration steps for MFE code itself have already been taken of within the existing `master-design-tokens` branches. 
  * @robrap I'm still planning to put together a similar, 2U-specific guide in our internal wiki (or a GH issue) when our dependencies are ready.

**Comment by adamstankiewicz** [comment]
In earlier discussions with @brian-smith-tcril (e.g., within Paragon Working Group), we've generally aligned towards mid-June. We/I feel **targeting June 18** should be adequate time to mitigate these remaining 2U issues, and formalize and coordinate a release plan with owning MFE squads.

**Comment by sarina** [comment]
I don't see a board field called "Review After Date" and I don't know how the June 18th merge date fits into a field called "Review After Date"

**Comment by brian-smith-tcril** [comment]
I updated this post to reflect the new template. I used May 22nd as the transition unblocked date because that was when the last of the `master-design-tokens` branches was created (https://github.com/openedx/frontend-app-publisher/pull/1003) - most of the branches had existed for a few weeks by then, but the `master-design-tokens` branches for `frontend-app-support-tools` and `frontend-app-publisher` were not created until May 21st and May 22nd respectively. The delay for those MFEs happened because they are not set up for development or testing without using a classic (non-tutor) devstack environment.

**Comment by robrap** [comment]
Thank you @adamstankiewicz and @brian-smith-tcril for clarifying everything. I wish you luck and hope everything goes smoothly.

@sarina: I updated the "Review After Date" field to June 18. I don't know why you don't see that. It's just a normal board field. I'm happy to have a call with you at some point if needed. Thanks.

**Comment by sarina** [comment]
I am looking at https://github.com/orgs/openedx/projects/9/views/6 and https://github.com/orgs/openedx/projects/9/views/9 but don't see this field. Could you point me at the view where it's visible?

**Comment by brian-smith-tcril** [comment]
I also wasn't sure where the "Review after Date" field was, I did a little digging and was only able to see it by going to the "DEPR Board" and then clicking on this issue from there so it opens in the side panel

```
https://github.com/orgs/openedx/projects/9/views/4?pane=issue&itemId=72278618&issue=openedx%7Cbrand-openedx%7C23
```

**Comment by feanil** [comment]
@brian-smith-tcril thanks for updating the various values in the ticket description.  This all looks good from my perspective and should be good to move forward without any issues.

**Comment by robrap** [comment]
On the sidebar of the issue, there are the Project fields. There is also an arrow that needs to be clicked to see other details. Here are the two views (closed and opened):

<img width="370" alt="Image" src="https://github.com/user-attachments/assets/eeef72db-d260-410a-a5f3-1f710e1adfa8" />

<img width="330" alt="Image" src="https://github.com/user-attachments/assets/179f0a23-3fa7-489b-9e4c-2b68ff602b34" />

And as noted, it appears on the board view.

**Comment by dianakhuang** [comment]
@brian-smith-tcril are there any updates on this?

**Comment by brian-smith-tcril** [comment]
Yes, this should definitely be in "Plan Completed."
